### DISCUSSION TITLE
How I got Laravel Precognition working

### DISCUSSION THREAD

**Discussion by jameswagoner** [lorisleiva/laravel-actions#249] [Show and tell]
I came across an issue about Precognition support when I ran into that issue recently and saw that there was no interest really in resolving it so I wanted to share how I solved it and possible get some feedback on it.

First I dug into what Precognition was doing when throwing its `Attempting to predict the outcome of the [{$class}::{$method}()] method but the method is not defined.` exception.

In the [PrecognitionControllerDispatcher](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Foundation/Routing/PrecognitionControllerDispatcher.php) class, you'll find `ensureMethodExists()` where it gets passed the `$controller` and `$method` that the `handle()` method was given.

With how Laravel Actions uses decorators to keep your action classes clean, `$controller` ends up being the controller decorator instead of the class you defined in the route.

Solution for me was to extend the [HandlePrecognitiveRequests](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Foundation/Http/Middleware/HandlePrecognitiveRequests.php) middleware and bind my own `PrecognitionControllerDispatcher` that passes the parsed controller class from the route definition.

```php
namespace Nexus\Precognition;

use Illuminate\Foundation\Routing\PrecognitionControllerDispatcher as BasePrecognitionControllerDispatcher;
use Illuminate\Routing\Route;

class PrecognitionControllerDispatcher extends BasePrecognitionControllerDispatcher
{
    public function dispatch(Route $route, $controller, $method): void
    {
        // In order to work with Laravel Actions
        // we need the controller class from the route action
        $controller = $route->action['controller'];

        $this->ensureMethodExists($controller, $method);

        $this->resolveParameters($route, $controller, $method);

        abort(204, headers: ['Precognition-Success' => 'true']);
    }
}
```

**Comment by AlexisVS** [comment]
@jameswagoner 
Can you explain more ?
